feat: Add generators module

gpranav162 opened the pull request:
Closes #432

senthurayyappan commented:
Hi @gpranav162, can you please fix all mypy errors (which are mostly due to missing type annotations)?

gpranav162 commented:
Hi @senthurayyappan, I've fixed all the mypy errors. Please let me know if any further changes are needed!

senthurayyappan commented:
Hey @gpranav162, great work! Can we modify the structure of signal generators to follow our update pattern so that they work well with other components? We want this module to be used for generating setpoints in the future. Here is an example use case (pseudo code):


```
from opensourceleg.actuators import DephyActpack
from opensourceleg.extras.generators import SineGenerator
from opensourceleg.utilities import SoftRealTimeLoop

if __name__ == "__main__":
    actpack = DephyActpack(...)
    sp = SineGenerator(...)
    clock = SoftRealTimeLoop(...)
    
    with actpack:
        actpack.set_control_mode(...)
        
        for t in clock:
            val = sp.update(t)
            actpack.set_motor_position(val)
```

senthurayyappan commented:
We could also support pre-generating outputs and have an iter pattern with a `__next__` method within the SignalGenerator class.

gpranav162 commented:
Hi @senthurayyappan, I've made the suggested changes:

1. Moved `CustomGenerator` class to the signal_generators module
2. Renamed `ImpulseGenerator` to `StepGenerator` (step_time set to zero outputs constant value) and removed the `ConstantGenerator` class
3. Refactored the base class to use the `update(t)` method instead of `__call__`
4. Added support for iteration over precomputed outputs with the `__next__` method

Please let me know if there’s anything else you’d like me to update!

senthurayyappan reviewed (approved):
Looks great! I added some modifications and fixes:

1. Expression evaluator class' `evaluate` method and made variables an array instead of a dict
2. Removed `__next__` iter patterns & `precompute` method as they didn't serve any purpose (I'm sorry for sending you down this road)
3. Made `generate` method public 
4. Removed uncommon arguments from ABC's `__init__`
5. Added gaussian and uniform noise types with configurable parameters
6. Removed `plot` method from ABC
7. Removed Noise and Composite generators
